TLSv1.3: Support early_data with external PSKs

mattcaswell opened the pull request:
Currently we can only send early_data with a ticket based PSK. This extends that capability to external PSKs too.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

kaduk commented on ssl/statem/extensions.c:
```diff
@@ -1242,6 +1242,13 @@ int tls_psk_do_binder(SSL *s, const EVP_MD *md, const unsigned char *msgstart,
+
+    if (external
+            && s->early_data_state == SSL_EARLY_DATA_CONNECTING
+            && s->session->ext.max_early_data == 0
```
What does the check s->session->ext.max_early_data == 0 do?
I guess it relates to having both an s->session and a s->psksession but don't fully undestand how that interacts yet.

mattcaswell replied:
It is theoretically possible for a client to provide *both* a ticket based session object *and* an external PSK based session object. In that case, if we are attempting to send early_data, then we prefer the ticket based session over the PSK. This check is about determining which one we are using. If `s->session->ext.max_early_data == 0` then the ticket based session does not support early_data.

kaduk replied:
I guess the spec does let you send multiple PSK+binders, sure.
I a little bit wish we had a better way to be consistent with "use the first PSK for early data" than to just hardcode everywhere that the ticket-based PSK goes first, but I don't have any concrete suggestions for achieving that.

kaduk commented on ssl/statem/extensions_clnt.c:
```diff
@@ -678,12 +678,41 @@ EXT_RETURN tls_construct_ctos_early_data(SSL *s, WPACKET *pkt,
+    const EVP_MD *handmd = NULL;
+
+    if (s->hello_retry_request)
+        handmd = ssl_handshake_md(s);
```
> Early data is not permitted after HelloRetryRequest.

Shouldn't we just return early here?
I guess that's inconsistent with moving the psk_use_session_cb() here from construct_psk but we still need to disable early data for HelloRetryRequest resposes, which I don't see happening here (am I missing it?).  We certainly can't rely on the callback detecting this case...

mattcaswell replied:
`s->early_data_state != SSL_EARLY_DATA_CONNECTING` in the event of an HRR so we will hit the `return EXT_NOT_SENT` below in that case. However it looks like HRR with early_data is screwed anyway. Fix in #3933. Probably I need to hold off with this PR until that has gone in, and then rebase this one.

kaduk commented on ssl/statem/extensions_clnt.c:
```diff
@@ -678,12 +678,41 @@ EXT_RETURN tls_construct_ctos_early_data(SSL *s, WPACKET *pkt,
+            return EXT_RETURN_FAIL;
+        }
+        s->psksession_id_len = idlen;
+    }
```
This feels like a natural place to add checks (required by section 2.3 of draft-21) that the psksess includes a TLS version number and cipher suite.  I guess the checks for ALPN/SNI might be a bit harder to do, but we can probably sanity-check that there was not something else configured already...

kaduk commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1603,10 +1621,20 @@ int tls_parse_stoc_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
+    /*
+     * If we used the external PSK for sending early_data then s->early_secret
+     * is already set up, so don't overwrite it. Otherwise we copy the
+     * early_secret across that we generated earlier.
```
Now I'm curious why the one generated "here" would be different from the one used for sending early data...

mattcaswell replied:
As noted above you can send *both* a ticket based session and an external PSK based session in the ClientHello. Only one can be used for sending early_data, but either can be selected by the server and they will have different early secret values. This code deals with the case where we sent early_data using the ticket based session, but the server selected the external PSK instead. In that scenario the early secret that we used for sending the early_data is no longer valid - we need to use the one associated with the external PSK instead.

kaduk commented on ssl/statem/extensions_srvr.c:
```diff
@@ -739,6 +739,7 @@ int tls_parse_ctos_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
             memcpy(sess->sid_ctx, s->sid_ctx, s->sid_ctx_length);
             sess->sid_ctx_length = s->sid_ctx_length;
             ext = 1;
+            s->ext.early_data_ok = 1;
```
Do we need to check for TLS version/cipher suite/ALPN/SNI here as well?

kaduk commented on ssl/tls13_enc.c:
```diff
@@ -404,6 +404,9 @@ int tls13_change_cipher_state(SSL *s, int which)
                 goto err;
             }
+
+            if (sslcipher == NULL && s->psksession != NULL)
```
Can we end up with a sslcipher in the "regular session" when a psksession is in use?
(If not, this could replace checking for a cipher suite stored in the psksession at the other two places I noted, but it's unclear that we want to wait until CCS to detect such errors.)

mattcaswell replied:
Hmmm. I think there could be a lurking bug here. I need to do some checks.

mattcaswell replied:
This bug is now fixed.

mattcaswell commented:
I have now significantly updated this PR. It turned out that adding the SNI/ALPN checks was quite involved. However they are now in, and I have rebased to resolve conflicts.

We now check on the client side that when you are attempting to write early data that your SNI/ALPN settings are consistent with your session (e.g. you are offering a set of ALPN protocols which includes the one set in your PSK). If not you get a fatal error. On the server side we have similar checks. However if they fail it just means that early_data will be rejected.

I have not, at the moment, added additional sanity checks on the client side after the server accepts or rejects early_data. Should I? For example a badly behaving server could select an ALPN protocol from the client's list which is not the same as the one from the original session BUT still accept early_data. I'm not sure what we should do in that case? Abort the connection?

davidben commented:
Yes. The client sanity check of the server selection is the important set of checks and is needed for correctness. Sanity-checking the session before using it is just an optimization for rare cases to not write early data that is guaranteed to be rejected.

If a server picks ALPN different from the one that was predicted in the session, the server should have declined 0-RTT. Having the protocol change partway through will confuse the caller, potentially introducing a bug in them. (On 0-RTT accept. Rejects allow all properties to change and need special consideration in the 0-RTT API for callers to be able to handle that right.)

mattcaswell commented:
Rebased and updated to add the late client sanity checking discussed above. Note this also includes the commit from #4181 - without which the tests fail. Ping @kaduk.

kaduk commented:
I guess I'll review #4181 first, then ... it's less daunting than this one :)

mattcaswell commented:
Hold off on reviewing this one for now then. Given the discussion in #4181 I'm going to need to make some tweaks here.

mattcaswell commented:
Following #4181 I have removed the late client detection of SNI inconsistencies from this PR. Also rebased. It's now ready for review (for real this time). Ping @kaduk.

mattcaswell commented on apps/s_client.c:
```diff
@@ -1888,6 +1888,26 @@ int s_client_main(int argc, char **argv)
         }
+        /* By default the SNI should be the same as was set in the session */
+        if (!noservername && servername == NULL)
+        {
```
Oops. Fixed.

mattcaswell commented on apps/s_client.c:
```diff
@@ -2600,8 +2620,10 @@ int s_client_main(int argc, char **argv)
+            && ((SSL_get0_session(con) != NULL
+                 && SSL_SESSION_get_max_early_data(SSL_get0_session(con)) > 0)
+                || (psksess != NULL
+                    && SSL_SESSION_get_max_early_data(psksess) > 0))) {
```
We don't silently ignore it. If we don't send it s_client writes the message "Early data was not sent"

mattcaswell commented on doc/man3/SSL_SESSION_get0_hostname.pod:
```diff
@@ -18,16 +29,32 @@ client when the session was created, or NULL if no value was sent.
+
+SSL_SESSION_set1_alpn_selected() sets the ALPN protocol for this session to the
+value in B<*alpn> which should be of length B<len> bytes. A copy of this value
+is taken.
```
Done

mattcaswell commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -59,6 +61,10 @@ determine if a session established with a server can be used to send early data.
 
+The function SSL_SESSION_set_max_early_data() sets the maximum number of early
+data bytes that can be sent for a session. This would typically be used when
+creating a PSK session file (see L<SSL_CTX_set_psk_use_session_callback(3)>).
```
Done

mattcaswell commented on ssl/record/ssl3_record_tls13.c:
```diff
@@ -58,7 +58,10 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
     if (s->early_data_state == SSL_EARLY_DATA_WRITING
             || s->early_data_state == SSL_EARLY_DATA_WRITE_RETRY) {
-        alg_enc = s->session->cipher->algorithm_enc;
+        if (s->session != NULL && s->session->ext.max_early_data > 0)
```
I added an ossl_assert() here too.

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -679,12 +679,85 @@ EXT_RETURN tls_construct_ctos_early_data(SSL *s, WPACKET *pkt,
+    if ((s->ext.alpn == NULL && edsess->ext.alpn_selected != NULL)) {
+        SSLerr(SSL_F_TLS_CONSTRUCT_CTOS_EARLY_DATA,
+               SSL_R_INCONSISTENT_EARLY_DATA_ALPN);
+        return EXT_RETURN_FAIL;
```
Why would it not be legal? The server may not accept our PSK and revert to a full handshake. In that case we may wish it to look at the ALPN list, even though we didn't use it for the early_data (which the server ignored).

kaduk replied:
Well, I did not spend too much time thinking about this question, so may well be wrong.
An assumption I had for my thinking was that we could be pretty sure the server supports ALPN negotiation to some extent (but I did not check whether that is actually required by the spec); similarly I was assuming that a server that supports ALPN and is presented with a list of protocols should pick one of them (and in fact, RFC 7301 seems to say that the server MUST send a fatal alert if ALPN negotiation fails).  So, if we've talked to this server before, and this server supports ALPN, we'd have to have been doing something extremely weird/suspect in order for the server's resumption state to not have an associated ALPN value.
But I consider this discussion to be a side note with respect to the substance of the pull request, and it can be resolved (or ignored) separately.

mattcaswell replied:
We are talking abut the scenario where the server previously issues a PSK ticket with no associated ALPN (which is what we are using for sending early_data). It does not seem unreasonable to me that, coming back some time later, we find that the server does now support ALPN. The configuration may have been updated, the server upgraded, or we may have been talking to a cluster of servers with varying capabilities.

kaduk replied:
Okay, I'll drop it.

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1338,6 +1401,24 @@ int tls_parse_stoc_alpn(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
 
+    if (s->session->ext.alpn_selected != NULL
+            && (s->session->ext.alpn_selected_len != len
+                || memcmp(s->session->ext.alpn_selected, s->s3->alpn_selected,
```
Yes! And the "&&" should be an "||"

mattcaswell commented on ssl/statem/extensions_srvr.c:
```diff
@@ -745,6 +748,8 @@ int tls_parse_ctos_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
             sess->sid_ctx_length = s->sid_ctx_length;
             ext = 1;
+            if (id == 0)
+                s->ext.early_data_ok = 1;
```
The psk extension is special. We parse it early in `ssl_get_prev_session()`.

kaduk replied:
Sigh, of course.  I just wasn't looking far enough away, I guess.

mattcaswell commented:
There is a trivial conflict in libssl.num so this will need a rebase before pushing. I haven't done that yet for ease of reviewing. New commit pushed to address feedback comments.

>  Did you put much thought into how to divvy things up between setting a servername for the server and not expecting a client extension for resumption?

I'm not sure I understand this question

kaduk commented:
> There is a trivial conflict in libssl.num so this will need a rebase before pushing. I haven't done that yet for ease of reviewing. New commit pushed to address feedback comments.

I was just noticing that; thanks for holding off on the rebase for my account.

> > Did you put much thought into how to divvy things up between setting a servername for the server and not expecting a client extension for resumption?

> I'm not sure I understand this question

Some of the tests added a serverflag "-servername localhost", but other tests just modified the list of extensions expected in the ClientHello [for resumption flows].  Either one of these is a valid way to adapt the tests to the new s_client behavior, but they would exercise different codepaths.  There is presumably some benefit in exercising a diverse set of codepaths, which means having to choose which testcases get which modification.  The question was attempting to ask whether the decision of which modification to use for any given test case was essentially random, or guided by some other factor.  (I'm now assuming that "arbitrary" or "essentially random" will be the answer.)

mattcaswell commented:
> Some of the tests added a serverflag "-servername localhost", but other tests just modified the list of extensions expected in the ClientHello [for resumption flows]. Either one of these is a valid way to adapt the tests to the new s_client behavior, but they would exercise different codepaths. There is presumably some benefit in exercising a diverse set of codepaths, which means having to choose which testcases get which modification. The question was attempting to ask whether the decision of which modification to use for any given test case was essentially random, or guided by some other factor. (I'm now assuming that "arbitrary" or "essentially random" will be the answer.)

I don't really recall my reasoning now. Probably mainly about what was the easiest way to do it :-)

mattcaswell commented:
CI was green, so pushed. Thanks for the review.
